File: Submariner/SBClientController.m
//
//  SBClient.m
//  Sub
//
//  Created by Rafaël Warnault on 14/05/11.
//
//  Copyright (c) 2011-2014, Rafaël Warnault
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of the Read-Write.fr nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "SBClientController.h"
#import "SBSubsonicParsingOperation.h"

#import "NSManagedObjectContext+Fetch.h"

#import "Submariner-Swift.h"


@interface SBClientController (Private)
- (void)unplayAllTracks;
- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type coverID:(NSString*) coverID searchResult:(SBSearchResult*)searchResult;
- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type coverID:(NSString*) coverID;
- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type searchResult:(SBSearchResult*)searchResult;
- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type;
@end



@implementation SBClientController


@synthesize managedObjectContext;
@synthesize delegate;
@synthesize connected;
@synthesize server;
@synthesize librarySection;
@synthesize remotePlaylistsSection;
@synthesize podcastsSection;
@synthesize radiosSection;
@synthesize searchsSection;
@synthesize home;
@synthesize library;
@synthesize isConnecting;
@synthesize queue;


#pragma mark -
#pragma mark LifeCycle

- (id)initWithManagedObjectContext:(NSManagedObjectContext *)context
{
    self = [super init];
    if (self) {
        parameters = [[NSMutableDictionary alloc] init];
        queue = [[NSOperationQueue alloc] init];
        [queue setMaxConcurrentOperationCount:1];
        managedObjectContext = context;
        connected = NO;
        isConnecting = NO;
        numberOfElements = 0;

    }
    
    return self;
}






#pragma mark -
#pragma mark Private




- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type coverID:(NSString*) coverID searchResult:(SBSearchResult*)searchResult {
	NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
	NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration];
	NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL: url];
	NSString *loginString = [NSString stringWithFormat: @"%@:%@", server.username, server.password];
	NSData *loginData = [loginString dataUsingEncoding: NSUTF8StringEncoding];
	NSString *base64login = [loginData base64EncodedStringWithOptions: 0];
	NSString *authHeader = [NSString stringWithFormat: @"Basic %@", base64login];
	configuration.HTTPAdditionalHeaders = @{@"Authorization": authHeader};
	NSURLSessionDataTask *httpTask = [session dataTaskWithRequest: request completionHandler:
		^(NSData *data, NSURLResponse *response, NSError *error) {
		dispatch_async(dispatch_get_main_queue(), ^{
		NSLog(@"Handling %@", url);
		if (error != nil) {
			NSLog(@"Error in requddestWithURL: %@", error);
			[NSApp presentError: error];
			return;
		}
		if (response == nil) {
			NSLog(@"No response in requestWithURL");
			return;
		}
		NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
		NSInteger statusCode = [httpResponse statusCode];
		NSLog(@"Status code is %ld", (long)statusCode);
		if (statusCode!= 200) {
			NSError *newError = nil;
			NSMutableDictionary *userInfo = [NSMutableDictionary dictionary];
			switch (statusCode) {
			case 400: [userInfo setValue:@"Bad request" forKey:NSLocalizedDescriptionKey];
				break;
			case 401: [userInfo setValue:@"Unauthorized" forKey:NSLocalizedDescriptionKey];
				break;
			case 402: [userInfo setValue:@"Payment Required" forKey:NSLocalizedDescriptionKey];
				break;
			case 403: [userInfo setValue:@"Forbidden" forKey:NSLocalizedDescriptionKey];
				break;
			case 404: [userInfo setValue:@"Not Found" forKey:NSLocalizedDescriptionKey];
				break;
			case 500: [userInfo setValue:@"Internal Error" forKey:NSLocalizedDescriptionKey];
				break;
			case 501: [userInfo setValue:@"Not Implemented" forKey:NSLocalizedDescriptionKey];
				break;
			case 502: [userInfo setValue:@"Service temporarily overloaded" forKey:NSLocalizedDescriptionKey];
				break;
			case 503: [userInfo setValue:@"Gateway timeout" forKey:NSLocalizedDescriptionKey];
				break;
			default: [userInfo setValue:@"Bad request" forKey:NSLocalizedDescriptionKey];
				break;
			}
			newError = [NSError errorWithDomain:NSPOSIXErrorDomain code: statusCode userInfo:userInfo];
			[NSApp presentError: newError];
			//[newError release];
			return;
		}

		SBSubsonicParsingOperation *operation = [[SBSubsonicParsingOperation alloc]
			initWithManagedObjectContext:self.managedObjectContext
			client:self
			requestType:type
			server:[self.server objectID]
			xml: data
            mimeType: response.MIMEType];
		if (type == SBSubsonicRequestGetCoverArt) {
			[operation setCurrentCoverID:coverID];
		} else if (type == SBSubsonicRequestSearch) {
			[operation setCurrentSearch:searchResult];
		} else {
			[[NSOperationQueue sharedServerQueue] cancelAllOperations];
		}
		[[NSOperationQueue sharedServerQueue] addOperation:operation];
		});
	}];
	[httpTask resume];
}

- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type coverID:(NSString*) coverID {
    NSParameterAssert(coverID != nil);
    [self requestWithURL:url requestType:type coverID:coverID searchResult:nil];
}

- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type searchResult:(SBSearchResult*)searchResult {
    NSParameterAssert(searchResult != nil);
    [self requestWithURL:url requestType:type coverID:nil searchResult:searchResult];
}

- (void)requestWithURL:(NSURL *)url requestType:(SBSubsonicRequestType)type {
    [self requestWithURL:url requestType:type coverID:nil searchResult:nil];
}

#pragma mark -
#pragma mark Request Messages

- (void)connectToServer:(SBServer *)aServer {
    [server getBaseParameters: parameters];
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/ping.view" parameters:parameters];
    [self requestWithURL:url requestType:SBSubsonicRequestPing];
}

- (void)getLicense {
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getLicense.view" parameters:parameters];
    [self requestWithURL:url requestType:SBSubsonicRequestGetLicense];
}


- (void)getIndexes {
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getIndexes.view" parameters:parameters];
    [self requestWithURL:url requestType:SBSubsonicRequestGetIndexes];
}

- (void)getIndexesSince:(NSDate *)date {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:[NSString stringWithFormat:@"%00.f", [date timeIntervalSince1970]] forKey:@"ifModifiedSince"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getIndexes.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestGetIndexes];
}


- (void)getAlbumsForArtist:(SBArtist *)artist {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:artist.id forKey:@"id"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getMusicDirectory.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestGetAlbumDirectory];
}


- (void)getAlbumsForArtistWithID:(SBArtistID *)artistID {
    
}


- (void)getCoverWithID:(NSString *)coverID {
    
    if(![NSThread isMainThread]) {
        [self performSelectorOnMainThread:@selector(getCoverWithID:) withObject:coverID waitUntilDone:YES];
        return;
    }

    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:coverID forKey:@"id"];
     [params setValue:[[NSUserDefaults standardUserDefaults] stringForKey:@"MaxCoverSize"] forKey:@"size"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getCoverArt.view" parameters:params];
    [self requestWithURL: url requestType: SBSubsonicRequestGetCoverArt coverID:coverID];
}


- (void)getTracksForAlbumID:(NSString *)albumID {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:albumID forKey:@"id"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getMusicDirectory.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestGetTrackDirectory];
}


- (void)getPlaylists {
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getPlaylists.view" parameters:parameters];
    [self requestWithURL:url requestType:SBSubsonicRequestGetPlaylists];
}

- (void)getPlaylist:(SBPlaylist *)playlist {
    // XXX: Why is this function different from the others?
    [server getBaseParameters: parameters];
    
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:playlist.id forKey:@"id"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getPlaylist.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestGetPlaylist];
}


- (void)getPodcasts {
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getPodcasts.view" parameters:parameters];
    [self requestWithURL:url requestType:SBSubsonicRequestGetPodcasts];
}


- (void)deletePlaylistWithID:(NSString *)playlistID {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:playlistID forKey:@"id"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/deletePlaylist.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestDeletePlaylist];
    
}

- (void)createPlaylistWithName:(NSString *)playlistName tracks:(NSArray *)tracks {
    // required parameters
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:playlistName forKey:@"name"];
    
    // compute params string (because obviously, dictionary doesn't support set of multiple same key)
    NSMutableString *paramString = [NSMutableString string];
    for (NSString *trackID in tracks) {
        [paramString appendFormat:@"&songId=%@", trackID];
    }
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/createPlaylist.view" parameters:params andParameterString:paramString];
    //NSLog(@"url : %@", url);
    [self requestWithURL:url requestType:SBSubsonicRequestCreatePlaylist];
}


- (void)updatePlaylistWithID:(NSString *)playlistID tracks:(NSArray *)tracks {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:playlistID forKey:@"playlistId"];
    
    // compute params string (because obviously, dictionary doesn't support set of multiple same key)
    NSMutableString *paramString = [NSMutableString string];
    for (NSString *trackID in tracks) {
        [paramString appendFormat:@"&songId=%@", trackID];
    }
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/createPlaylist.view" parameters:params andParameterString:paramString];
    [self requestWithURL:url requestType:SBSubsonicRequestCreatePlaylist];
}


- (void)getAlbumListForType:(SBSubsonicRequestType)type {
    
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    if(type == SBSubsonicRequestGetAlbumListRandom) {
        [params setValue:@"random" forKey:@"type"];
        
    } else if(type == SBSubsonicRequestGetAlbumListNewest) {
        [params setValue:@"newest" forKey:@"type"];
        
    } else if(type == SBSubsonicRequestGetAlbumListFrequent) {
        [params setValue:@"frequent" forKey:@"type"];
        
    } else if(type == SBSubsonicRequestGetAlbumListHighest) {
        [params setValue:@"highest" forKey:@"type"];
        
    } else if(type == SBSubsonicRequestGetAlbumListRecent) {
        [params setValue:@"recent" forKey:@"type"];
    }
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getAlbumList.view" parameters:params];
    [self requestWithURL:url requestType:type];
}

- (void)getNowPlaying {
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getNowPlaying.view" parameters:parameters];
    [self requestWithURL:url requestType:SBSubsonicRequestGetNowPlaying];   
}


- (void)getUserWithName:(NSString *)username {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:username forKey:@"username"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/getUser.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestGetUser];  
}


- (void)search:(NSString *)query {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:query forKey:@"query"];
    [params setValue:@"100" forKey:@"songCount"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/search2.view" parameters:params];
    SBSearchResult *searchResult = [[SBSearchResult alloc] initWithQuery:query];
    [self requestWithURL:url requestType: SBSubsonicRequestSearch searchResult:searchResult];
}


- (void)setRating:(NSInteger)rating forID:(NSString *)anID {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:[NSString stringWithFormat:@"%ld", rating] forKey:@"rating"];
    [params setValue:anID forKey:@"id"];
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/setRating.view " parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestSetRating];  
}


- (void)scrobble:(NSString *)anID {
    NSMutableDictionary *params = [NSMutableDictionary dictionaryWithDictionary:parameters];
    [params setValue:anID forKey:@"id"];
    unsigned long long currentTimeMS = (unsigned long long)([[NSDate date] timeIntervalSince1970] * 1000);
    [params setValue: [NSString stringWithFormat:@"%lld", currentTimeMS] forKey:@"time"];
    // XXX: Set submissions?
    
    NSURL *url = [NSURL URLWithString:server.url command:@"rest/scrobble.view" parameters:params];
    [self requestWithURL:url requestType:SBSubsonicRequestScrobble];
}



@end

File: Submariner/SBClientController.swift
//
//  SBClientController.swift
//  Submariner
//
//  Created by Calvin Buckley on 2023-06-13.
//  Copyright © 2023 Submariner Developers. All rights reserved.
//

import Cocoa
import os

fileprivate let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "SBClientController")

@objc class SBClientController: NSObject {
    let managedObjectContext: NSManagedObjectContext
    let server: SBServer // XXX: weak?
    
    var parameters: [String: String] = [:]
    
    @objc init(managedObjectContext: NSManagedObjectContext, server: SBServer) {
        self.managedObjectContext = managedObjectContext
        self.server = server
        super.init()
    }
    
    // #MARK: - HTTP Requests
    
    private func request(url: URL, type: SBSubsonicParsingOperation.RequestType, customization: ((SBSubsonicParsingOperation) -> Void)? = nil) {
        let config = URLSessionConfiguration.default
        let session = URLSession(configuration: config)
        let request = URLRequest(url: url)
        // No auth header needed since we just pass them over query string
        
        let task = session.dataTask(with: request) { data, response, error in
            // sensitive because &p= contains user password
            logger.info("Handling URL \(url, privacy: .sensitive)")
            logger.info("\tAPI endpoint \(url.path, privacy: .public)")
            
            if let error = error {
                DispatchQueue.main.async {
                    NSApp.presentError(error)
                }
                return
            } else if let response = response as? HTTPURLResponse {
                logger.info("\tStatus code is \(response.statusCode)")
                // Note that Subsonic and Navidrome return app-level error bodies in HTTP 200
                // HTTP 404s are used for unsupported features in Subsonic and Navidrome,
                // but OC Music uses code 70 Subsonic responses
                if response.statusCode == 404 {
                    self.server.markNotSupported(feature: type)
                    return
                } else if response.statusCode == 429 {
                    // Newer versions of Navidrome back getCoverArt w/ third-party APIs.
                    // As such, it rate limits API requests that can invoke them.
                    // Instead of bothering the user, retry the request later.

                    // Retry-After is seconds or a specific date
                    let retryAfter = response.value(forHTTPHeaderField: "Retry-After")
                    logger.info("Retrying w/ Retry-After value \(retryAfter ?? "<nil>")")

                    if let retryAfter = retryAfter,
                       let specificDate = retryAfter.dateTimeFromHTTP() {
                        _ = Timer(fire: specificDate, interval: 0, repeats: false) { timer in
                            self.request(url: url, type: type, customization: customization)
                            timer.invalidate()
                        }
                    } else {
                        // handle if Retry-After is valid, invalid, or missing
                        let seconds = TimeInterval(retryAfter ?? "5") ?? 5
                        _ = Timer(timeInterval: seconds, repeats: false) { timer in
                            self.request(url: url, type: type, customization: customization)
                            timer.invalidate()
                        }
                    }
                    return
                } else if response.statusCode != 200 {
                    let message = "HTTP \(response.statusCode) for \(url.path)"
                    let userInfo = [NSLocalizedDescriptionKey: message]
                    // XXX: Right domain?
                    let error = NSError(domain: NSURLErrorDomain, code: response.statusCode, userInfo: userInfo)
                    DispatchQueue.main.async {
                        NSApp.presentError(error)
                    }
                    return
                }
                
                if let operation = SBSubsonicParsingOperation(managedObjectContext: self.managedObjectContext,
                                                              requestType: type,
                                                              server: self.server.objectID,
                                                              xml: data,
                                                              mimeType: response.mimeType) {
                    if let customization = customization {
                        customization(operation)
                    }
                    OperationQueue.sharedServerQueue.addOperation(operation)
                }
            }
        }
        task.resume()
    }
    
    // HACK: eat the nils here if some stupid nil happens,
    // without needing to change logic in real func or add ifs to all the callsites
    private func request(url: URL?, type: SBSubsonicParsingOperation.RequestType, customization: ((SBSubsonicParsingOperation) -> Void)? = nil) {
        if let url = url {
            request(url: url, type: type, customization: customization)
        } else {
            logger.error("URL was nil for request \(String(describing: type)), the server URL likely needs to be reset")
        }
    }
    
    // #MARK: - Request Messages
    
    @objc func connect(server: SBServer) {
        parameters = parameters.mergedCopyFrom(dictionary: server.getBaseParameters())
        let url = URL.URLWith(string: server.url, command: "rest/ping.view", parameters: parameters)
        request(url: url, type: .ping)
    }
    
    @objc func getLicense() {
        let url = URL.URLWith(string: server.url, command: "rest/getLicense.view", parameters: parameters)
        request(url: url, type: .getLicense)
    }
    
    func getArtists() {
        let url = URL.URLWith(string: server.url, command: "rest/getArtists.view", parameters: parameters)
        request(url: url, type: .getArtists)
    }
    
    func get(artist: SBArtist) {
        var params = parameters
        if artist.itemId == nil {
            // can happen because of now playing/search

            return
        }
        params["id"] = artist.itemId
        
        let url = URL.URLWith(string: server.url, command: "rest/getArtist.view", parameters: params)
        request(url: url, type: .getArtist) { operation in
            operation.currentArtistID = artist.itemId
        }
    }
    
    func getCover(id: String, for albumID: String? = nil) {
        var params = parameters
        params["id"] = id
        
        let url = URL.URLWith(string: server.url, command: "rest/getCoverArt.view", parameters: params)
        request(url: url, type: .getCoverArt) { operation in
            operation.currentCoverID = id
            operation.currentAlbumID = albumID
        }
    }
    
    func getTrack(trackID: String) {
        var params = parameters
        params["id"] = trackID
        
        let url = URL.URLWith(string: server.url, command: "rest/getSong.view", parameters: params)
        request(url: url, type: .getTrack)
    }
    
    func get(album: SBAlbum) {
        var params = parameters
        params["id"] = album.itemId
        
        let url = URL.URLWith(string: server.url, command: "rest/getAlbum.view", parameters: params)
        request(url: url, type: .getAlbum) { operation in
            operation.currentAlbumID = album.itemId
        }
    }
    
    @objc func getPlaylists() {
        let url = URL.URLWith(string: server.url, command: "rest/getPlaylists.view", parameters: parameters)
        request(url: url, type: .getPlaylists)
    }
    
    @objc func getPlaylist(_ playlist: SBPlaylist) {
        var params = parameters
        params["id"] = playlist.itemId
        
        let url = URL.URLWith(string: server.url, command: "rest/getPlaylist.view", parameters: params)
        request(url: url, type: .getPlaylist) { operation in
            operation.currentPlaylistID = playlist.itemId
        }
    }
    
    @objc func getPodcasts() {
        let url = URL.URLWith(string: server.url, command: "rest/getPodcasts.view", parameters: parameters)
        request(url: url, type: .getPodcasts)
    }
    
    @objc(deletePlaylistWithID:) func deletePlaylist(id: String) {
        var params = parameters
        params["id"] = id
        
        let url = URL.URLWith(string: server.url, command: "rest/deletePlaylist.view", parameters: params)
        request(url: url, type: .deletePlaylist) { operation in
            operation.currentPlaylistID = id
        }
    }
    
    @objc(createPlaylistWithName:tracks:) func createPlaylist(name: String, tracks: [SBTrack]) {
        var params = parameters
        params["name"] = name
        
        // XXX: DRY this with update
        let allParams = params.map { (k, v) in  URLQueryItem(name: k, value: v) } +
            tracks.map { track in URLQueryItem(name: "songId", value: track.itemId) }
        
        let url = URL.URLWith(string: server.url, command: "rest/createPlaylist.view", queryItems: allParams)
        request(url: url, type: .createPlaylist)
    }
    
    @objc(updatePlaylistWithID:tracks:) func updatePlaylist(playlistID: String, tracks: [SBTrack]) {
        var params = parameters
        params["playlistId"] = playlistID
        
        let allParams = params.map { (k, v) in  URLQueryItem(name: k, value: v) } +
            tracks.map { track in URLQueryItem(name: "songId", value: track.itemId) }
        
        let url = URL.URLWith(string: server.url, command: "rest/createPlaylist.view", queryItems: allParams)
        request(url: url, type: .updatePlaylist) { operation in
            operation.currentPlaylistID = playlistID
        }
    }
    
    func updatePlaylist(ID: String,
                        name: String? = nil,
                        comment: String? = nil,
                        isPublic: Bool? = nil,
                        appending: [SBTrack]? = nil,
                        removing: [Int]? = nil) {
        var params = parameters
        params["playlistId"] = ID
        if let name = name {
            params["name"] = name
        }
        if let comment = comment {
            params["comment"] = comment
        }
        if let isPublic = isPublic {
            params["public"] = "\(isPublic)"
        }
        
        let allParams = params.map { (k, v) in  URLQueryItem(name: k, value: v) } +
            (appending?.map { track in URLQueryItem(name: "songIdToAdd", value: track.itemId) } ?? []) +
            (removing?.map { index in URLQueryItem(name: "songIndexToRemove", value: "\(index)") } ?? [])
        
        let url = URL.URLWith(string: server.url, command: "rest/updatePlaylist.view", queryItems: allParams)
        request(url: url, type: .updatePlaylist) { operation in
            operation.currentPlaylistID = ID
        }
        
    }
    
    @objc(getAlbumListForType:) func getAlbumList(type: SBSubsonicParsingOperation.RequestType) {
        var params = parameters
        switch type {
        case .getAlbumListRandom:
            params["type"] = "random"
        case .getAlbumListNewest:
            params["type"] = "newest"
        case .getAlbumListFrequent:
            params["type"] = "frequent"
        case .getAlbumListHighest:
            params["type"] = "highest"
        case .getAlbumListRecent:
            params["type"] = "recent"
        default:
            logger.error("getAlbumList type: unrecognized")
            abort()
        }
        
        let url = URL.URLWith(string: server.url, command: "rest/getAlbumList2.view", parameters: params)
        request(url: url, type: type)
    }
    
    @objc func getNowPlaying() {
        let url = URL.URLWith(string: server.url, command: "rest/getNowPlaying.view", parameters: parameters)
        request(url: url, type: .getNowPlaying)
    }
    
    @objc(getUserWithName:) func getUser(username: String) {
        var params = parameters
        params["username"] = username
        
        let url = URL.URLWith(string: server.url, command: "rest/getUser.view", parameters: params)
        request(url: url, type: .getUser)
    }
    
    @objc func search(_ query: String) {
        var params = parameters
        params["query"] = query
        params["songCount"] = "100" // XXX: Configurable? Pagination?
        
        let url = URL.URLWith(string: server.url, command: "rest/search3.view", parameters: params)
        request(url: url, type: .search) { operation in
            operation.currentSearch = SBSearchResult(query: query)
        }
    }
    
    @objc(setRating:forID:) func setRating(_ rating: Int, id: String) {
        var params = parameters
        params["rating"] = String(rating)
        params["id"] = id
        
        let url = URL.URLWith(string: server.url, command: "rest/setRating.view", parameters: params)
        request(url: url, type: .setRating)
    }
    
    @objc func scrobble(id: String) {
        var params = parameters
        params["id"] = id
        let currentTimeMS = Int64(Date().timeIntervalSince1970 * 1000)
        params["time"] = String(currentTimeMS)
        
        let url = URL.URLWith(string: server.url, command: "rest/scrobble.view", parameters: params)
        request(url: url, type: .scrobble)
    }
    
    @objc func scanLibrary() {
        let url = URL.URLWith(string: server.url, command: "rest/startScan.view", parameters: parameters)
        request(url: url, type: .getNowPlaying)
    }
    
    @objc func getScanStatus() {
        let url = URL.URLWith(string: server.url, command: "rest/getScanStatus.view", parameters: parameters)
        request(url: url, type: .getNowPlaying)
    }
}

extension Dictionary {
    func mergedCopyFrom(dictionary: Dictionary) -> Dictionary {
        var new = self
        for (k, v) in dictionary {
            new[k] = v
        }
        return new
    }
}
